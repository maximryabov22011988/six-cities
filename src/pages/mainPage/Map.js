import React, { Component } from 'react';
import PropTypes from 'prop-types';
import leaflet from 'leaflet';

const mapStyles = {
  top: '15px',
  height: '794px'
};

const DEFAULT_MAP_ZOOM = 12;

const LEAFLET_LAYER = {
  URL_TEMPLATE:
    'https://{s}.basemaps.cartocdn.com/rastertiles/voyager/{z}/{x}/{y}{r}.png',
  OPTIONS: {
    attribution:
      '&copy; <a href="https://www.openstreetmap.org/copyright">OpenStreetMap</a> contributors &copy; <a href="https://carto.com/attributions">CARTO</a>'
  }
};

const propTypes = {
  offers: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.string,
      title: PropTypes.string,
      image: PropTypes.string,
      price: PropTypes.number,
      type: PropTypes.oneOf(['Apartment', 'Private room']),
      rating: PropTypes.number,
      isPremium: PropTypes.bool,
      isBookmark: PropTypes.bool,
      location: PropTypes.arrayOf(PropTypes.number)
    })
  ).isRequired,
  city: PropTypes.arrayOf(PropTypes.number).isRequired
};

class Map extends Component {
  constructor(props) {
    super(props);

    this.settings = {
      center: props.city,
      zoom: DEFAULT_MAP_ZOOM,
      zoomControl: false,
      marker: true
    };
  }

  componentDidMount() {
    this.initCard();
  }

  get pinIcon() {
    return leaflet.icon({
      iconUrl: 'img/pin.svg',
      iconSize: [30, 40]
    });
  }

  addPins() {
    const { offers } = this.props;
    const icon = this.pinIcon;
    offers.forEach(({ location, title }) => {
      leaflet.marker(location, { icon, title }).addTo(this.offerMap);
    });
  }

  initCard() {
    const { settings } = this;
    const { city } = this.props;

    this.offerMap = leaflet.map('map', settings);
    this.offerMap.setView(city, DEFAULT_MAP_ZOOM);
    leaflet
      .tileLayer(LEAFLET_LAYER.URL_TEMPLATE, LEAFLET_LAYER.OPTIONS)
      .addTo(this.offerMap);

    this.addPins();
  }

  render() {
    return <div id="map" style={mapStyles} />;
  }
}

Map.propTypes = propTypes;

export default Map;
