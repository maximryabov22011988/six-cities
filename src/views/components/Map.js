import React, { Component } from 'react';
import PropTypes from 'prop-types';
import Leaflet from 'leaflet';

const radiusInKm = 3;

const circleStyles = {
  color: '#FF9000',
  fillColor: '#FF9000',
  fillOpacity: 0.1,
  radius: radiusInKm * 1000,
};

const mapStyles = {
  fixed: {
    top: 0,
    zIndex: 100,
    width: '1144px',
    height: '580px',
    marginLeft: 'auto',
    marginRight: 'auto',
  },
  flexible: {
    top: 0,
    height: '100%',
    zIndex: 100,
  },
};

const mapContainerId = 'map';

const LEAFLET_LAYER = {
  URL_TEMPLATE: 'https://{s}.basemaps.cartocdn.com/rastertiles/voyager/{z}/{x}/{y}{r}.png',
  OPTIONS: {
    attribution:
      '&copy; <a href="https://www.openstreetmap.org/copyright">OpenStreetMap</a> contributors &copy; <a href="https://carto.com/attributions">CARTO</a>',
  },
};

const propTypes = {
  className: PropTypes.string,
  currentOffer: PropTypes.shape({
    bedrooms: PropTypes.number,
    city: PropTypes.shape({
      name: PropTypes.string,
      location: PropTypes.object,
    }),
    description: PropTypes.string,
    goods: PropTypes.arrayOf(PropTypes.string),
    host: PropTypes.shape({
      avatar_url: PropTypes.string,
      id: PropTypes.number,
      is_pro: PropTypes.bool,
      name: PropTypes.string,
    }),
    id: PropTypes.number,
    images: PropTypes.arrayOf(PropTypes.string),
    is_favorite: PropTypes.bool,
    is_premium: PropTypes.bool,
    location: PropTypes.shape({
      latitude: PropTypes.number,
      longitude: PropTypes.number,
      zoom: PropTypes.number,
    }),
    max_adults: PropTypes.number,
    preview_image: PropTypes.string,
    price: PropTypes.number,
    rating: PropTypes.number,
    title: PropTypes.string,
    type: PropTypes.string,
  }),
  offers: PropTypes.arrayOf(
    PropTypes.shape({
      bedrooms: PropTypes.number,
      city: PropTypes.shape({
        name: PropTypes.string,
        location: PropTypes.object,
      }),
      description: PropTypes.string,
      goods: PropTypes.arrayOf(PropTypes.string),
      host: PropTypes.shape({
        avatar_url: PropTypes.string,
        id: PropTypes.number,
        is_pro: PropTypes.bool,
        name: PropTypes.string,
      }),
      id: PropTypes.number,
      images: PropTypes.arrayOf(PropTypes.string),
      is_favorite: PropTypes.bool,
      is_premium: PropTypes.bool,
      location: PropTypes.shape({
        latitude: PropTypes.number,
        longitude: PropTypes.number,
        zoom: PropTypes.number,
      }),
      max_adults: PropTypes.number,
      preview_image: PropTypes.string,
      price: PropTypes.number,
      rating: PropTypes.number,
      title: PropTypes.string,
      type: PropTypes.string,
    }),
  ),
};

class Map extends Component {
  componentDidMount() {
    const { currentOffer, offers } = this.props;

    const { latitude, longitude, zoom } = offers ? offers[0].city.location : currentOffer.location;

    const settings = {
      center: [latitude, longitude],
      zoom,
      zoomControl: false,
    };

    this.offerMap = Leaflet.map(mapContainerId, settings);
    this.offerMap.setView([latitude, longitude], zoom);
    Leaflet.tileLayer(LEAFLET_LAYER.URL_TEMPLATE, LEAFLET_LAYER.OPTIONS).addTo(this.offerMap);
    this.pinGroup = Leaflet.layerGroup().addTo(this.offerMap);
    this.addPins(this.pinGroup);
  }

  componentDidUpdate() {
    const { currentOffer, offers } = this.props;

    const { latitude, longitude, zoom } = offers ? offers[0].city.location : currentOffer.location;

    if (this.offerMap) {
      this.offerMap.setView([latitude, longitude], zoom);
      this.pinGroup.clearLayers();
      this.addPins(this.pinGroup);
    }
  }

  get pinIcon() {
    return Leaflet.icon({
      iconUrl: 'img/pin.svg',
      iconSize: [27, 39],
      iconAnchor: [13, 39],
    });
  }

  get activePinIcon() {
    return Leaflet.icon({
      iconUrl: 'img/pin-active.svg',
      iconSize: [27, 39],
      iconAnchor: [13, 39],
    });
  }

  addPins(group) {
    const { currentOffer, offers } = this.props;
    const icon = this.pinIcon;
    const activeIcon = this.activePinIcon;

    if (currentOffer) {
      const {
        location: { latitude, longitude },
      } = currentOffer;
      Leaflet.marker([latitude, longitude], { icon: activeIcon }).addTo(group);
      Leaflet.circle([latitude, longitude], circleStyles).addTo(group);
    }

    offers.forEach(({ location: { latitude, longitude } }) => {
      Leaflet.marker([latitude, longitude], { icon }).addTo(group);
    });
  }

  render() {
    const { className, fixed } = this.props;
    return <div className={className} id={mapContainerId} style={fixed ? mapStyles.fixed : mapStyles.flexible} />;
  }
}

Map.propTypes = propTypes;

export default Map;
