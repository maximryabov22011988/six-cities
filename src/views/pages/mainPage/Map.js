import React, { Component } from 'react';
// import PropTypes from 'prop-types';
import Leaflet from 'leaflet';

const mapStyles = {
  top: '15px',
  height: '794px'
};

const LEAFLET_LAYER = {
  URL_TEMPLATE:
    'https://{s}.basemaps.cartocdn.com/rastertiles/voyager/{z}/{x}/{y}{r}.png',
  OPTIONS: {
    attribution:
      '&copy; <a href="https://www.openstreetmap.org/copyright">OpenStreetMap</a> contributors &copy; <a href="https://carto.com/attributions">CARTO</a>'
  }
};

// const propTypes = {
//   offers: PropTypes.arrayOf(
//     PropTypes.shape({
//       id: PropTypes.string,
//       title: PropTypes.string,
//       image: PropTypes.string,
//       price: PropTypes.number,
//       type: PropTypes.oneOf(['Apartment', 'Private room']),
//       rating: PropTypes.number,
//       isPremium: PropTypes.bool,
//       isBookmark: PropTypes.bool,
//       location: PropTypes.arrayOf(PropTypes.number)
//     })
//   ).isRequired
// };

class Map extends Component {
  constructor(props) {
    super(props);

    this.leaflet = Leaflet;

    this.icon = this.leaflet.icon({
      iconUrl: 'img/pin.svg',
      iconSize: [27, 39],
      iconAnchor: [13, 39]
    });
  }

  componentDidMount() {
    const { offers } = this.props;

    if (offers.length > 0) {
      this.map = this.leaflet.map(
        'map',
        {
          center: [
            offers[0].city.location.latitude,
            offers[0].city.location.longitude
          ],
          zoom: offers[0].city.location.zoom,
          zoomControl: false,
          layers: new this.leaflet.TileLayer(
            LEAFLET_LAYER.URL_TEMPLATE,
            LEAFLET_LAYER.OPTIONS
          )
        },
        100
      );

      this.markers = this.leaflet.layerGroup().addTo(this.map);
      this.addMarkers(offers, this.markers);
    }
  }

  componentDidUpdate() {
    const { offers } = this.props;

    if (this.map) {
      this.map.setView(
        [offers[0].city.location.latitude, offers[0].city.location.longitude],
        offers[0].city.location.zoom
      );
      this.markers.clearLayers();
      this.addMarkers(offers, this.markers);
    }
  }

  addMarkers(offers, group) {
    offers.map(offer => {
      this.leaflet
        .marker([offer.location.latitude, offer.location.longitude], {
          icon: this.icon
        })
        .addTo(group);
    });
  }

  render() {
    return <div id="map" style={mapStyles} />;
  }
}

// Map.propTypes = propTypes;

export default Map;
