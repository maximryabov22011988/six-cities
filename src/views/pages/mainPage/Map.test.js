import React from 'react';
import renderer from 'react-test-renderer';

import Map from './Map';

const offersMock = [
  {
    id: 'card-1',
    title: 'Beautiful &amp; luxurious apartment at great location',
    image: 'img/apartment-01.jpg',
    price: 120,
    type: 'Apartment',
    rating: 4.6,
    isPremium: true,
    isBookmark: false,
    location: [52.3909553943508, 4.85309666406198]
  }
];

const city = [52.38333, 4.9];

it(`Map correctly render`, () => {
  Map.prototype.componentDidMount = jest.fn();
  const map = renderer.create(<Map offers={offersMock} city={city} />).toJSON();

  expect(map).toMatchSnapshot();
});
