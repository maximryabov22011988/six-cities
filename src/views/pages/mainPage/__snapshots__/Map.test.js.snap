// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Map correctly render 1`] = `
<div
  id="map"
  style={
    Object {
      "height": "792px",
      "top": 0,
      "zIndex": 100,
    }
  }
/>
`;
